In [ ]:
import tensorflow
import tensorflow as tf
from tensorflow import keras
from keras import Sequential
from keras.layers import Dense,Flatten,Conv2D,MaxPool2D,Dropout
from keras.applications.vgg16 import VGG16
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.preprocessing import image
import warnings
from tensorflow.keras.applications.mobilenet_v2 import preprocess_input
from google.colab import files
import numpy as np
warnings.filterwarnings('ignore')
from tensorflow.keras.preprocessing import image
from tensorflow.keras.applications.vgg16 import preprocess_input

In [ ]:
input_shape_non_gen=(224,224,3)

In [ ]:
conv_base = VGG16(
    weights='imagenet',
    include_top = False,
    input_shape=input_shape_non_gen
)

58889256/58889256 ━━━━━━━━━━━━━━━━━━━━ 4s 0us/step


In [ ]:
conv_base.summary()

Model: "vgg16"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ input_layer (InputLayer)             │ (None, 224, 224, 3)         │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block1_conv1 (Conv2D)                │ (None, 224, 224, 64)        │           1,792 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block1_conv2 (Conv2D)                │ (None, 224, 224, 64)        │          36,928 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block1_pool (MaxPooling2D)           │ (None, 112, 112, 64)        │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block2_conv1 (Conv2D)                │ (None, 112, 112, 128)       │          73,856 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block2_conv2 (Conv2D)                │ (None, 112, 112, 128)       │         147,584 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block2_pool (MaxPooling2D)           │ (None, 56, 56, 128)         │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block3_conv1 (Conv2D)                │ (None, 56, 56, 256)         │         295,168 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block3_conv2 (Conv2D)                │ (None, 56, 56, 256)         │         590,080 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block3_conv3 (Conv2D)                │ (None, 56, 56, 256)         │         590,080 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block3_pool (MaxPooling2D)           │ (None, 28, 28, 256)         │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block4_conv1 (Conv2D)                │ (None, 28, 28, 512)         │       1,180,160 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block4_conv2 (Conv2D)                │ (None, 28, 28, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block4_conv3 (Conv2D)                │ (None, 28, 28, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block4_pool (MaxPooling2D)           │ (None, 14, 14, 512)         │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block5_conv1 (Conv2D)                │ (None, 14, 14, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block5_conv2 (Conv2D)                │ (None, 14, 14, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block5_conv3 (Conv2D)                │ (None, 14, 14, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block5_pool (MaxPooling2D)           │ (None, 7, 7, 512)           │               0 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 14,714,688 (56.13 MB)

 Trainable params: 14,714,688 (56.13 MB)

 Non-trainable params: 0 (0.00 B)

In [ ]:
conv_base.trainable = True
set_trainable = False

for layer in conv_base.layers:
  if layer.name == 'block5_conv1':
    set_trainable = True
  if set_trainable:
    layer.trainable = True
  else:
    layer.trainable = False

for layer in conv_base.layers:
  print(layer.name,layer.trainable)

input_layer False
block1_conv1 False
block1_conv2 False
block1_pool False
block2_conv1 False
block2_conv2 False
block2_pool False
block3_conv1 False
block3_conv2 False
block3_conv3 False
block3_pool False
block4_conv1 False
block4_conv2 False
block4_conv3 False
block4_pool False
block5_conv1 True
block5_conv2 True
block5_conv3 True
block5_pool True


In [ ]:
model = Sequential()
model.add(conv_base)  # Add the VGG16 convolutional base

# Add new layers for your task
model.add(Flatten())
model.add(Dense(256, activation='relu'))
model.add(Dropout(0.4))  # Optional dropout to prevent overfitting
model.add(Dense(5, activation='softmax'))

In [ ]:
train_ds = keras.utils.image_dataset_from_directory(
    directory='/content/drive/MyDrive/dataset/train',  # Replace with your training directory
    labels='inferred',
    label_mode='categorical',  # Adjust for your classification problem
    image_size=(224, 224),  # Specify image size here
    batch_size=32,
    shuffle=True,
    seed=42)

Found 1210 files belonging to 5 classes.


In [ ]:
validation_ds = keras.utils.image_dataset_from_directory(
    directory='/content/drive/MyDrive/dataset/valid',  # Replace with your training directory
    image_size=(224, 224),  # Match your image size - Changed from target_size to image_size
    batch_size=32,
    label_mode='categorical',  # Adjust for your classification problem - Changed from class_mode to label_mode
    shuffle=False,
    seed=42
# For reproducibility
)

Found 289 files belonging to 5 classes.


In [ ]:
test_ds = tf.keras.utils.image_dataset_from_directory(
    directory='/content/drive/MyDrive/dataset/test',
    labels='inferred',
    image_size=(224, 224),
    batch_size=32,
    label_mode='categorical',
    shuffle=False,
    seed=42
)

Found 30 files belonging to 5 classes.


In [ ]:
def process(image,label):
    image = tensorflow.cast(image/255. ,tensorflow.float32)
    return image,label


In [ ]:
train_ds = train_ds.map(process)
validation_ds = validation_ds.map(process)
test_ds = test_ds.map(process)

In [ ]:
model.compile(
    loss='categorical_crossentropy',
    optimizer=keras.optimizers.Adam(learning_rate=1e-5),
    metrics=['accuracy']
)

In [ ]:
##for non-generator

early_stopping = keras.callbacks.EarlyStopping(patience=5, restore_best_weights=True)

history = model.fit(train_ds,epochs=15,batch_size=32,validation_data=validation_ds,callbacks=[early_stopping]
)

Epoch 1/15
38/38 ━━━━━━━━━━━━━━━━━━━━ 499s 13s/step - accuracy: 0.2838 - loss: 1.7181 - val_accuracy: 0.7543 - val_loss: 0.9646
Epoch 2/15
38/38 ━━━━━━━━━━━━━━━━━━━━ 11s 288ms/step - accuracy: 0.6669 - loss: 0.9166 - val_accuracy: 0.7993 - val_loss: 0.5734
Epoch 3/15
38/38 ━━━━━━━━━━━━━━━━━━━━ 19s 251ms/step - accuracy: 0.8427 - loss: 0.4752 - val_accuracy: 0.8616 - val_loss: 0.4311
Epoch 4/15
38/38 ━━━━━━━━━━━━━━━━━━━━ 10s 270ms/step - accuracy: 0.9135 - loss: 0.3200 - val_accuracy: 0.8720 - val_loss: 0.3673
Epoch 5/15
38/38 ━━━━━━━━━━━━━━━━━━━━ 10s 264ms/step - accuracy: 0.9428 - loss: 0.2069 - val_accuracy: 0.8858 - val_loss: 0.3252
Epoch 6/15
38/38 ━━━━━━━━━━━━━━━━━━━━ 10s 250ms/step - accuracy: 0.9725 - loss: 0.1377 - val_accuracy: 0.8962 - val_loss: 0.2966
Epoch 7/15
38/38 ━━━━━━━━━━━━━━━━━━━━ 11s 279ms/step - accuracy: 0.9570 - loss: 0.1439 - val_accuracy: 0.8962 - val_loss: 0.2981
Epoch 8/15
38/38 ━━━━━━━━━━━━━━━━━━━━ 20s 262ms/step - accuracy: 0.9851 - loss: 0.0901 - val_accur

In [ ]:
model.evaluate(test_ds)

1/1 ━━━━━━━━━━━━━━━━━━━━ 26s 26s/step - accuracy: 0.9333 - loss: 0.1743


[0.17431659996509552, 0.9333333373069763]

In [ ]:
# for loading model for pretrained
model.save('/content/drive/MyDrive/dataset/weather classification cnn pre-trained.h5')

In [ ]:
# Create file uploader widget
uploaded = files.upload()
# Get the uploaded image
for fn in uploaded.keys():
  img = image.load_img(fn, target_size=(224, 224))
  x = image.img_to_array(img)
  x = np.expand_dims(x, axis=0)
  x = preprocess_input(x)


Saving rainy5.jpg to rainy5.jpg


In [ ]:
predictions = model.predict(x)
# Determine the predicted class
predicted_class_index = np.argmax(predictions)
class_mapping = {
    0: 'sunny',
    1: 'cloudy',
    2: 'rainy',
    3: 'sunrise',
    4: 'foggy'
}
# Use class_mapping directly instead of weather_classes
predicted_weather = class_mapping[predicted_class_index]
print('Predicted weather:', predicted_weather)
print('Predicted weather:', predicted_weather)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 492ms/step
Predicted weather: rainy
Predicted weather: rainy


# Summary

## Evaluation

*   use vgg16 as pretrained model
*   It's val_accuracy: 0.9170
*   model evaluation on test data set 0.93
*   It takes alot of time without gpu.
*   If you don't have alot of data than use pretrained model but you need have gpu other wise it will take alot of time 4-6hour for training










